File: src/main/java/no/nav/veilarboppfolging/service/ArenaOppfolgingService.java
package no.nav.veilarboppfolging.service;

import lombok.extern.slf4j.Slf4j;
import no.nav.common.client.aktoroppslag.AktorOppslagClient;
import no.nav.common.client.norg2.Norg2Client;
import no.nav.common.types.identer.AktorId;
import no.nav.common.types.identer.Fnr;
import no.nav.pto_schema.enums.arena.Formidlingsgruppe;
import no.nav.pto_schema.enums.arena.Kvalifiseringsgruppe;
import no.nav.veilarboppfolging.NotFoundException;
import no.nav.veilarboppfolging.client.veilarbarena.ArenaOppfolgingTilstand;
import no.nav.veilarboppfolging.client.veilarbarena.VeilarbArenaOppfolging;
import no.nav.veilarboppfolging.client.veilarbarena.VeilarbarenaClient;
import no.nav.veilarboppfolging.controller.response.OppfolgingEnhetMedVeilederResponse;
import no.nav.veilarboppfolging.repository.OppfolgingsStatusRepository;
import no.nav.veilarboppfolging.repository.VeilederTilordningerRepository;
import no.nav.veilarboppfolging.repository.entity.OppfolgingEntity;
import no.nav.veilarboppfolging.utils.ArenaUtils;
import no.nav.veilarboppfolging.utils.EnumUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;

import static no.nav.veilarboppfolging.utils.SecureLog.secureLog;

@Slf4j
@Service
public class ArenaOppfolgingService {

    // Bruker AktorregisterClient istedenfor authService for å unngå sirkulær avhengighet
    private final AktorOppslagClient aktorOppslagClient;

    private final VeilarbarenaClient veilarbarenaClient;

    private final OppfolgingsStatusRepository oppfolgingsStatusRepository;

    private final AuthService authService;

    private final Norg2Client norg2Client;

    private final VeilederTilordningerRepository veilederTilordningerRepository;

    @Autowired
    public ArenaOppfolgingService(
            AktorOppslagClient aktorOppslagClient,
            VeilarbarenaClient veilarbarenaClient,
            OppfolgingsStatusRepository oppfolgingsStatusRepository,
            AuthService authService,
            Norg2Client norg2Client,
            VeilederTilordningerRepository veilederTilordningerRepository
    ) {
        this.aktorOppslagClient = aktorOppslagClient;
        this.veilarbarenaClient = veilarbarenaClient;
        this.oppfolgingsStatusRepository = oppfolgingsStatusRepository;
        this.authService = authService;
        this.norg2Client = norg2Client;
        this.veilederTilordningerRepository = veilederTilordningerRepository;
    }

    // Bruker endepunktet i veilarbarena som henter fra database som er synket med Arena (har et delay på et par min)
    public Optional<VeilarbArenaOppfolging> hentOppfolgingFraVeilarbarena(Fnr fnr) {
        return veilarbarenaClient.hentOppfolgingsbruker(fnr);
    }

    // Bruker endepunktet i veilarbarena som henter direkte fra Arena
    public Optional<ArenaOppfolgingTilstand> hentOppfolgingTilstandDirekteFraArena(Fnr fnr) {
        return veilarbarenaClient.getArenaOppfolgingsstatus(fnr).map(ArenaOppfolgingTilstand::fraArenaOppfolging);
    }

    public Optional<ArenaOppfolgingTilstand> hentOppfolgingTilstand(Fnr fnr) {
        AktorId aktorId = aktorOppslagClient.hentAktorId(fnr);

        Optional<ArenaOppfolgingTilstand> maybeArenaOppfolging = veilarbarenaClient.hentOppfolgingsbruker(fnr)
                .map(ArenaOppfolgingTilstand::fraArenaBruker);

        // TODO: Kan hente erUnderOppfolging gjennom OppfolgingService
        Optional<OppfolgingEntity> oppfolging = oppfolgingsStatusRepository.hentOppfolging(aktorId);
        boolean erUnderOppfolging = oppfolging.map(OppfolgingEntity::isUnderOppfolging).orElse(false);

        boolean erUnderOppfolgingIVeilarbarena = maybeArenaOppfolging
                .map(o -> ArenaUtils.erUnderOppfolging(EnumUtils.valueOf(Formidlingsgruppe.class, o.getFormidlingsgruppe()) , EnumUtils.valueOf(Kvalifiseringsgruppe.class, o.getServicegruppe())))
                .orElse(false);

        if (erUnderOppfolgingIVeilarbarena != erUnderOppfolging) {
            var oppfolgingTilstand = hentOppfolgingTilstandDirekteFraArena(fnr);

            maybeArenaOppfolging.ifPresent(arenaOppfolging -> {
                log.info(
                        "Differanse mellom oppfolging fra veilarbarena og direkte fra Arena."
                                + " veilarbarena.formidlingsgruppe={} veilarbarena.servicegruppe={}"
                                + " arena.formidlingsgruppe={} arena.servicegruppe={}",
                        arenaOppfolging.getFormidlingsgruppe(),
                        arenaOppfolging.getServicegruppe(),
                        oppfolgingTilstand.map(ArenaOppfolgingTilstand::getFormidlingsgruppe).orElse(null),
                        oppfolgingTilstand.map(ArenaOppfolgingTilstand::getServicegruppe).orElse(null)
                );
            });

            return oppfolgingTilstand;
        }

        return maybeArenaOppfolging;
    }

    public OppfolgingEnhetMedVeilederResponse getOppfolginsstatus(Fnr fnr) {

        VeilarbArenaOppfolging veilarbArenaOppfolging = veilarbarenaClient.hentOppfolgingsbruker(fnr)
                .orElseThrow(() -> new NotFoundException("Bruker ikke funnet"));

        OppfolgingEnhetMedVeilederResponse oppfolgingEnhetMedVeileder = new OppfolgingEnhetMedVeilederResponse()
                .setServicegruppe(veilarbArenaOppfolging.getKvalifiseringsgruppekode())
                .setFormidlingsgruppe(veilarbArenaOppfolging.getFormidlingsgruppekode())
                .setOppfolgingsenhet(hentEnhet(veilarbArenaOppfolging.getNav_kontor()))
                .setHovedmaalkode(veilarbArenaOppfolging.getHovedmaalkode());

        if (authService.erInternBruker()) {
            AktorId brukersAktoerId = authService.getAktorIdOrThrow(fnr);
            secureLog.info("Henter tilordning for bruker med aktørId {}", brukersAktoerId);
            String veilederIdent = veilederTilordningerRepository.hentTilordningForAktoer(brukersAktoerId);
            oppfolgingEnhetMedVeileder.setVeilederId(veilederIdent);
        }

        return oppfolgingEnhetMedVeileder;
    }

    private OppfolgingEnhetMedVeilederResponse.Oppfolgingsenhet hentEnhet(String enhetId) {
        String enhetNavn = "";

        try {
            enhetNavn = norg2Client.hentEnhet(enhetId).getNavn();
        } catch (Exception e) {
            log.warn("Fant ikke navn på enhet", e);
        }

        return new OppfolgingEnhetMedVeilederResponse.Oppfolgingsenhet(enhetNavn, enhetId);
    }


}

File: src/main/kotlin/no/nav/veilarboppfolging/oppfolgingsbruker/arena/ArenaOppfolgingService.kt
package no.nav.veilarboppfolging.oppfolgingsbruker.arena

import java.time.LocalDate
import java.util.Optional
import lombok.extern.slf4j.Slf4j
import no.nav.common.client.aktoroppslag.AktorOppslagClient
import no.nav.common.client.norg2.Norg2Client
import no.nav.common.types.identer.EnhetId
import no.nav.common.types.identer.Fnr
import no.nav.pto_schema.enums.arena.Formidlingsgruppe
import no.nav.pto_schema.enums.arena.Hovedmaal
import no.nav.pto_schema.enums.arena.Kvalifiseringsgruppe
import no.nav.veilarboppfolging.FantIkkeBrukerIArenaException
import no.nav.veilarboppfolging.client.veilarbarena.ArenaOppfolgingTilstand
import no.nav.veilarboppfolging.client.veilarbarena.ArenaOppfolginsBrukerOppslagResult
import no.nav.veilarboppfolging.client.veilarbarena.RegistrerIArenaResult
import no.nav.veilarboppfolging.client.veilarbarena.VeilarbArenaOppfolgingsBruker
import no.nav.veilarboppfolging.client.veilarbarena.VeilarbArenaOppfolgingsStatus
import no.nav.veilarboppfolging.client.veilarbarena.VeilarbarenaClient
import no.nav.veilarboppfolging.oppfolgingsbruker.arena.OppfolgingEnhetMedVeilederResponse.Oppfolgingsenhet
import no.nav.veilarboppfolging.repository.OppfolgingsStatusRepository
import no.nav.veilarboppfolging.service.AuthService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import kotlin.jvm.optionals.getOrNull

@Slf4j
@Service
class ArenaOppfolgingService @Autowired constructor(
    // Bruker AktorregisterClient istedenfor authService for å unngå sirkulær avhengighet
    private val aktorOppslagClient: AktorOppslagClient,
    private val veilarbarenaClient: VeilarbarenaClient,
    private val oppfolgingsStatusRepository: OppfolgingsStatusRepository,
    private val authService: AuthService,
    private val norg2Client: Norg2Client,
) {
    private val log = LoggerFactory.getLogger(ArenaOppfolgingService::class.java)

    fun kanEnkeltReaktiveres(fnr: Fnr): Optional<Boolean> {
        return veilarbarenaClient.getArenaOppfolgingsstatus(fnr)
            .flatMap { Optional.ofNullable(it.kanEnkeltReaktiveres) }
    }

    fun brukerErIservIArena(fnr: Fnr): Boolean {
        return veilarbarenaClient.getArenaOppfolgingsstatus(fnr)
            .map { it.formidlingsgruppe == Formidlingsgruppe.ISERV.name }
            .orElse(false) // Nei hvis bruker ikke finnes i arena eller ikke får svar fra arena
    }

    /**
     *  Brukes kun hvis man trenger [VeilarbArenaOppfolgingsStatus.kanEnkeltReaktiveres] , dette feltet kommer ikke på topic og kan derfor ikke caches i lokalt
     *  @see no.nav.veilarboppfolging.client.veilarbarenaVeilarbArenaOppfolgingsStatus#kanEnkeltReaktiveres
     *  */
    fun hentArenaOppfolgingsStatus(fnr: Fnr): Optional<VeilarbArenaOppfolgingsStatus> {
        return veilarbarenaClient.getArenaOppfolgingsstatus(fnr)
    }

    /**
     *  Henter arena-oppfolgingstilstand lokalt med fallback til arena /oppfolgingsbruker.
     *  Eksponerer bare felter som er felles for /oppfolgingsbruker og /oppfolgingsstatus:
     *  - ikke [VeilarbArenaOppfolgingsStatus.kanEnkeltReaktiveres]
     *  - ikke [VeilarbArenaOppfolgingsBruker.hovedmaalkode]
     *  */
    fun hentArenaOppfolgingTilstand(fnr: Fnr): ArenaOppfolgingTilstandOppslagResult {
        val aktorId = aktorOppslagClient.hentAktorId(fnr)
        val oppfolging = oppfolgingsStatusRepository.hentOppfolging(aktorId)
        val localArenaOppfolging = oppfolging.getOrNull()?.localArenaOppfolging?.getOrNull()
        if (localArenaOppfolging != null) {
            return ArenaOppfolgingTilstandOppslagResult.Success(ArenaOppfolgingTilstand.fraLocalArenaOppfolging(localArenaOppfolging))
        }
        val result = veilarbarenaClient.hentOppfolgingsbruker(fnr)
        return when (result) {
            is ArenaOppfolginsBrukerOppslagResult.Fail -> ArenaOppfolgingTilstandOppslagResult.Fail(
                "Feil ved henting av oppfolgingsbruker fra arena",
                result.reason
            )
            is ArenaOppfolginsBrukerOppslagResult.NotFound -> ArenaOppfolgingTilstandOppslagResult.NotFound()
            is ArenaOppfolginsBrukerOppslagResult.Success -> ArenaOppfolgingTilstandOppslagResult.Success(
                ArenaOppfolgingTilstand.fraArenaBruker(result.oppfolgingsBruker)
            )
        }
    }

    fun hentIservDatoOgFormidlingsGruppe(fnr: Fnr): IservDatoOgFormidlingsGruppe? {
        val arenaOppfolingTilstandResult = hentArenaOppfolgingTilstand(fnr)
        return when(arenaOppfolingTilstandResult) {
            is ArenaOppfolgingTilstandOppslagResult.Fail, is ArenaOppfolgingTilstandOppslagResult.NotFound  -> null
            is ArenaOppfolgingTilstandOppslagResult.Success -> {
                val iservDato = arenaOppfolingTilstandResult.arenaOppfolingTilstand.inaktiveringsdato
                val formidlingsgruppe = arenaOppfolingTilstandResult.arenaOppfolingTilstand.formidlingsgruppe?.let {
                    Formidlingsgruppe.valueOf(it)
                }
                IservDatoOgFormidlingsGruppe(iservDato, formidlingsgruppe)
            }
        }
    }

    /* Egentlig ikke oppfolgingsstatus men oppfølgingsbruker men endepunktet heter /oppfolgingsstatus */
    fun hentArenaOppfolginsstatusMedHovedmaal(fnr: Fnr): GetOppfolginsstatusResult {
        val aktorId = authService.getAktorIdOrThrow(fnr)

        val lokaltLagretOppfolging = oppfolgingsStatusRepository.hentOppfolging(aktorId)
        val localArenaOppfolging = lokaltLagretOppfolging.flatMap { it.localArenaOppfolging }
        val oppfolgingsEnhet: EnhetId? = lokaltLagretOppfolging.map { it.oppfolgingsEnhet }.orElse(null)

        val oppfolgingsData: OppfolgingsData = when {
            (localArenaOppfolging.isPresent) -> {
                OppfolgingsData(
                    localArenaOppfolging.get().kvalifiseringsgruppe,
                    localArenaOppfolging.get().formidlingsgruppe,
                    localArenaOppfolging.get().hovedmaal
                )
            }
            else -> {
                val oppfolgingsbrukerOppslag = veilarbarenaClient.hentOppfolgingsbruker(fnr)
                val veilarbArenaOppfolging = when(oppfolgingsbrukerOppslag) {
                    is ArenaOppfolginsBrukerOppslagResult.Success -> oppfolgingsbrukerOppslag.oppfolgingsBruker
                    is ArenaOppfolginsBrukerOppslagResult.NotFound, is ArenaOppfolginsBrukerOppslagResult.Fail -> return GetOppfolginsstatusFailure(FantIkkeBrukerIArenaException())
                }
                OppfolgingsData(
                    Kvalifiseringsgruppe.valueOf(veilarbArenaOppfolging.kvalifiseringsgruppekode),
                    Formidlingsgruppe.valueOf(veilarbArenaOppfolging.formidlingsgruppekode),
                    veilarbArenaOppfolging.hovedmaalkode?.let { Hovedmaal.valueOf(it) }
                )
            }
        }

        val veilederIdent =
            if (authService.erInternBruker()) lokaltLagretOppfolging.map { it.veilederId }.orElse(null) else null
        return GetOppfolginsstatusSuccess(
            OppfolgingEnhetMedVeilederResponse(
                oppfolgingsenhet = oppfolgingsEnhet?.let { hentEnhet(it) },
                veilederId = veilederIdent,
                formidlingsgruppe = oppfolgingsData.formidlingsgruppe.name,
                servicegruppe = oppfolgingsData.kvalifiseringsgruppe.name,
                hovedmaalkode = oppfolgingsData.hovedmaal?.name,
            )
        )
    }

    fun registrerIkkeArbeidssoker(fnr: Fnr): RegistrerIArenaResult {
        return veilarbarenaClient.registrerIkkeArbeidsoker(fnr)
    }

    private fun hentEnhet(enhetId: EnhetId?): Oppfolgingsenhet? {
        if (enhetId == null) return null
        try {
            val enhetNavn = norg2Client.hentEnhet(enhetId.get()).getNavn()
            return Oppfolgingsenhet(navn = enhetNavn, enhetId = enhetId.get())
        } catch (e: Exception) {
            log.warn("Fant ikke navn på enhet", e)
            return Oppfolgingsenhet("", enhetId.get())
        }
    }
}

sealed class ArenaOppfolgingTilstandOppslagResult {
    class Success(val arenaOppfolingTilstand: ArenaOppfolgingTilstand) : ArenaOppfolgingTilstandOppslagResult()
    class NotFound : ArenaOppfolgingTilstandOppslagResult()
    class Fail(val message: String, val reason: Throwable) : ArenaOppfolgingTilstandOppslagResult()
}

data class OppfolgingsData(
    var kvalifiseringsgruppe: Kvalifiseringsgruppe,
    var formidlingsgruppe: Formidlingsgruppe,
    var hovedmaal: Hovedmaal?
)

data class IservDatoOgFormidlingsGruppe(
    val iservDato: LocalDate?,
    val formidlingsGruppe: Formidlingsgruppe?
)